[Discussion] Address Core Contributor portion of NumFOCUS feedback (#1280)
arviz-devs/arviz: GOVERNANCE.md

In other words, Council Members are first and foremost Core Contributors, but only
when needed they can collectively make decisions for the health of the project.

### Council Decision Making Process
By and large we expect the decisions in ArviZ to be made _ad hoc_ and require little formal
coordination and with the community at large. However for controversial proposals and
new core contributors the council may need to intervene to make the final decision
in a group vote.

#### Call for a vote
Core contributors can call for a vote on to resolve a target issue they feel has
been stale for too long and informal consensus appears unlikely. For a voting
to be called, the target issue must be at least 2 months old.

To do so, they have to open a proposal issue ticket labeled "Council Vote".
The proposal issue should contain a link to the target issue and a proposal
on how to resolve it. Proposals should include a statement making clear what it
means to "agree" or to "disagree".

Before voting starts, at least 3 days will be left for Core
Contributors to raise doubts about proposal _phrasing_, no extra discussion will take

> place in the proposal issue. Proposal issues should be locked from creation

Reviewer: Discussion _should have already taken place_ in the target (original) issue and Slack.

GitHub issues can be locked so that only people with merge privileges can comment. I locked #1260 for example to prevent stray comments. As all core contributors will have merge privileges, this effectively allows only core members to comment with phrasing complains and council members to vote. This is to prevent comments on the issue that are not part of the voting process. These should go in the target issue which would remain unlocked.

> place in the proposal issue. Proposal issues should be locked from creation

Reviewer 2: Ok got it, the current wording is clear then 👍

to prevent attracting discussion from people not familiar with the decision
process.

#### Voting process

* Each council member will vote either Yes, No, or Neutral.

>   * "No" votes must list the reasons for disagreement.

Reviewer: We can add that too, I'm not sure how useful it would be to have yes reason, the proposal should already expose the reasons for the agreement option and the yes reasons are bound to be the same.

>   * "No" votes must list the reasons for disagreement.

Reviewer 2: > the proposal should already expose the reasons for the agreement option

Agreed. I'm just saying that because the wording here could be interpreted as intimidating for dissenters and I think we should strive to allow for constructive disagreement. Maybe amend to something like: "Just as proposals should expose the reasons for agreement, "No" votes should explain why they disagree." Is that better? Just to show it's not easier to vote "Yes" than "No".

> the yes reasons are bound to be the same

Here I'd disagree: I think people often agree on common proposals but for different reasons -- but that's another debate 😜

Council Members will have the responsibility of
* Removing members, including Council Members, if they are in violation of the Code of Conduct

> * Make decisions when regular community discussion does not produce consensus on an issue

Reviewer 3: ```suggestion
* Making decisions when regular community discussion does not produce consensus on an issue
```

  in a reasonable time frame.

> * Make decisions about strategic collaborations with other organizations or individuals.
> * Make decisions about the overall scope, vision and direction of the project.

Reviewer 3: ```suggestion
* Making decisions about strategic collaborations with other organizations or individuals.
* Making decisions about the overall scope, vision and direction of the project.
```

* Developing funding sources
* Deciding how to disburse funds with consultation from Core Contributors

The council may choose to delegate these responsibilities to sub-committees. If so, Council members must update this document to make the delegation clear.

> Note that individual council member does not have the power to unilaterally wield these

Reviewer 3: ```suggestion
Note that an individual council member does not have the power to unilaterally wield these
```

responsibilities, but the council as a whole must jointly make these decisions.
In other words, Council Members are first and foremost Core Contributors, but only
when needed they can collectively make decisions for the health of the project.

### Council Decision Making Process
By and large we expect the decisions in ArviZ to be made _ad hoc_ and require little formal

> coordination and with the community at large. However for controversial proposals and

Reviewer 3: ```suggestion
coordination and with the community at large. However, for controversial proposals and
```

> new core contributors the council may need to intervene to make the final decision

Reviewer 3: ```suggestion
new Core Contributors the council may need to intervene to make the final decision
```

in a group vote.

#### Call for a vote

> Core contributors can call for a vote to resolve a target issue they feel has

Reviewer 3: ```suggestion
Core Contributors can call for a vote to resolve a target issue they feel has
```

> been stale for too long and informal consensus appears unlikely. For a voting

Reviewer 3: ```suggestion
been stale for too long and for which informal consensus appears unlikely. For a vote
```

to be called, the target issue must be at least 2 months old.

To do so, they have to open a proposal issue ticket labeled "Council Vote".
The proposal issue should contain a link to the target issue and a proposal
on how to resolve it. Proposals should include a statement making clear what it
means to "agree" or to "disagree".

Before voting starts, at least 3 days will be left for Core
Contributors to raise doubts about the proposal's _phrasing_, no extra discussion will take
place in the proposal issue. Proposal issues should be locked from creation
to prevent attracting discussion from people not familiar with the decision
process.

#### Voting process

> * Each council member will vote either Yes, No, or Neutral.

Reviewer 3: ```suggestion
* Each Council Member will vote either "Yes", "No", or "Neutral".
```

> * It is recommended that all council members expose their reasons when voting.

Reviewer 3: ```suggestion
* It is recommended that all Council Members expose their reasons when voting.
```

>   "No" votes however _must_ list the reasons for disagreement. Any "no" vote

Reviewer 3: ```suggestion
  "No" votes, however, _must_ list the reasons for disagreement. Any "No" vote
```

>   with no reason listed will be considered a "neutral" vote.

Reviewer: How about now @AlexAndorra ? It is true that the reasons behind a yes vote may not be the same as the reasons listed in the proposal, but I still think that requiring no votes to list their reasons is not making it more difficult to vote no. I'll try to explain better why and provide some example.

Proposals are submitted by core contributors who are trusted members of the community to the point they have merge privileges and have had to go through a selection process mediated by the council. We should therefore expect them to have ArviZ best interests in mind. 

Therefore, I feel that voting "yes" without listing any reasons (or even voting yes because the proposal aligns with your idea) does not do much harm to the project nor the community. Voting no on the other hand goes along way towards knocking down the proposal. Knocking down proposals can make the people who proposed not good enough, that they do not belong, not worthy of being a core contributor... _especially if the proposal is knocked down without explaining why_

>   with no reason listed will be considered a "neutral" vote.

Reviewer 3: ```suggestion
  with no reason listed will be considered a "Neutral" vote.
```

> * An absence of vote is considered as neutral.

Reviewer 3: ```suggestion
* An absence of vote is considered as "Neutral".
```

* Voting will remain open for at least 3 days.

> * For the proposal to pass, at least 60% of the council must vote Yes, and no more

Reviewer 3: ```suggestion
* For the proposal to pass, at least 60% of the council must vote "Yes", and no more
```

>   than 20% can vote No.

Reviewer 3: ```suggestion
  than 20% can vote "No".
```

For decisions about the project the Council will perform it directly on the
proposal issue.

> For decisions about people, such as Core Contributors or ejecting folks, the Council

Reviewer 3: "folks" is very informal for this document. Also, what does "ejecting" mean here?

Reviewer: I think ejecting here refers to the `Ejecting Core Contributors` section. Maybe we could reword as:

    For decisions about people, such as electing or ejecting Core Contributors, the Council

Reviewer: Also, this should probably go in another PR but is the ejecting for core contributors only? What about single time contributors (or people with no relation to ArviZ) being disrespective in issues and breaking CoC?

Reviewer 3: I'm fine with that change.

Note that individual council member does not have the power to unilaterally wield these responsibilities, but the council as a whole must jointly make these decisions. In other words, Council Members are first and foremost Core Contributors, but only when needed they can collectively make decisions for the health of the project.

### Council Decision Making Process

> By and large we expect the decisions in ArviZ to be made ad hoc and require formal coordination and with the community at large. However for conterversial proposals and new core contributors the council will make the final decision in a group vote.

Reviewer 2: Don't you mean "and require _little_ formal coordination"?

> * Each council member will vote either Yes, No, or Neutral
>   * A non vote is considered nuetral
> * Voting will remain open for at least 3 days
> * To pass 60% of the council must vote Yes, and no more than 25% can vote No

Reviewer 2: ```suggestion
* Each council member will vote either Yes, No, or Neutral.
* An absence of vote is considered as neutral.
* Voting will remain open for at least 3 days.
* For the proposal to pass, at least 60% of the council must vote Yes, and no more than 20% can vote No.
```

* Must be core contributor for at least one year

#### Nominations
* Nominations are taken over a public GitHub issue ticket over the course of 2 weeks.
* Only Core Contributors may nominate folks
* Self Nominations are allowed
* At the conclusion of the 2 weeks, the list of nominations is posted on the ticket and this ticket is closed.

#### Election Process

* Voting occurs over a period of at least 1 week, at the conclusion of the nominations.
  Voting is blind and mediated by either an application or a third party like NumFOCUS.
  Each voter can vote zero or more times, once per each candidate. As this is not about

>   ranking but about capabilities, voters vote on a yes/neutral/no basis

Reviewer: I modified to a yes/neutral/no basis, not the yes/no type of voting we did for the first election. Now the following question arises.

> Do we interpret yes=+1, neutral=0 and no=-1 and leave the point below unchanged?
> Or do we use the same 2 threshold pattern?

In the second option, candidate face "only themselves" in the sense that the results of other candidates do not affect them being chosen -> there is no possibility for ties. We may have to restrict the min and max number of council members though, otherwise we could face a situation where nobody gets chosen.

Moreover, in any of the 2 cases we need some rules on how to modify the size of the council.

Reviewer: I propose having the following steps:

* yes/neutral/no voting
   - if the number of accepted candidates is in the interval [4, 13] (or whatever) election results are definitive (unless challenged)
   - otherwise we do yes=1, neutral=0, no=-1 and rank
* If we get here it means we are in the +1,0,-1 alternative so we can keep points from 2 onwards as they are now to break ties.

Reviewer 2: > - yes/neutral/no voting
        - if the number of accepted candidates is in the interval [4, 13] (or whatever) election results are definitive (unless challenged)

This looks good to me, with Majority Judgment to break potential ties. This is simple, focused on capabilities instead of ranking, and is basically an extension of what we did for the first election, extended with a "neutral" option

Reviewer: @AlexAndorra I am not sure I understand the comment, mainly due to the quote of the 1a step. Assuming we chose 4, 13 as thresholds for max and min number of council members:
* if there are 15 candidates and 12 get yes>=60% and no<=20%, all 12 are directly elected as council members. 
* if instead all 15 get yes>=60% and no<=20%, they would then be ranked by interpreting yes=1, neutral=0, no=-1 and the first 13 would be elected council members
   - if there were ties (e.g. 12, 13 and 14 candidates have all a score of 4) the 2nd election would be triggered exactly as it is written now:

> In the event of a tie there will be a runoff election for the tied candidates. To avoid
  further ties and discriminate more among the tied candidates, this vote will be held
  by [Majority Judgment](https://en.wikipedia.org/wiki/Majority_judgment) (MJ): for each
  candidate, voters judge their suitability for office as either Excellent, Very Good,
  Good, Acceptable, Poor, or Reject. Multiple candidates may be given the same grade by
  a voter. The candidate with the highest median grade is the winner.

  many candidates were confirmed, the 10 candidates with higher rank are
  elected. If not enough candidates were chosen, the 4 candidates with higher
  rank are elected.
* In the event of a tie there will be a runoff election for the tied candidates. To avoid
  further ties and discriminate more among the tied candidates, this vote will be held
  by [Majority Judgment](https://en.wikipedia.org/wiki/Majority_judgment) (MJ): for each
  candidate, voters judge their suitability for office as either Excellent, Very Good,
  Good, Acceptable, Poor, or Reject. Multiple candidates may be given the same grade by
  a voter. The candidate with the highest median grade is the winner.
* If more than one candidate has the same highest median-grade, the MJ winner is
  discovered by removing (one-by-one) any grades equal in value to the shared median
  grade from each tied candidate's total. This is repeated until only one of the
  previously tied candidates is currently found to have the highest median-grade.
* If ties are still present after this second round, the winner will be chosen at random.
  Each person tied will pick a number between 0 and 100, and a random integer will be

>   generated from random.org. The person with the closet circular distance will be selected.

Reviewer 4: closest

>   generated from random.org. The person with the closest circular distance will be selected.

Reviewer 3: Who picks the integer? How will this not be easy to game?

Reviewer 3: Define circular distance?

Reviewer 3: Explicitly state the random integer will be in [0,100]?

Reviewer: I would say the same person (aka numfocus representative) who handled the other two previous voting does. I think a difficult to game process would be all candidates send their number to numfocus and they generate the number. The process is repeated as many times as necessary (the same number may be chosen leading to yet another tie).

About circular distance, how about `min(|𝑎−𝑏|,100−|𝑎−𝑏|)` (taken from https://math.stackexchange.com/questions/1951658/calculate-distance-between-two-values-in-a-cycle) and have generated numbers be in [0, 99] (or use 101 and [0, 100])?

Reviewer 3: How about [1,100] (which is the default for the front page generator at random.org) and `min(|𝑎−𝑏|,100−|𝑎−𝑏|)`? This seems to do the right thing.

### Core Contributors
Core Contributors are those who have provided consistent and meaningful contributions to ArviZ.
These can be, but are not limited to, code contributions, community contributions, tutorial
development etc. Core Contributors will be given the ability to manage the ArviZ github
repository, including code merges to master. This does not necessarily mean Core Contributors
must submit code, but more so signifies trust with the project as a whole.

#### Core Contributor Responsibilities
* Enforce code of conduct
* Maintain a check against Council

#### Core Contributor Nominations and Confirmation Process

> Current Core Contributors can nominate candidates to become Core Contributors. If nominated candidates accept their nomination (explicit comment approving nomination on the GitHub issue or "thumbsup" emoji on the same issue), then they can be considered by the Council: on the first of the month following a nomination, the Council will vote on each nominee, where greater than 50% of the votes means a confirmation. In the event of a tie, a second election will occur, where all core contributors can vote. If 50% or greater of the community confirm, the nominee will be accepted.

Reviewer: The second sentence assumes nominations are public over GitHub issues if I am correctly understanding the proposal but it is not clear nor explicit. Would we open an issue per contributor to be added? Or would there be a core contributor nomination issue where we'd add new comments every time someone is nominated? 

Regarding the voting results, how public should it be? Do council members vote publicly or privately? Is the whole result published or only yes/no (probably better not yet than no)?

Reviewer: Should also be updated after deciding the voting style

Reviewer 2: You're right -- Didn't realize this was for Core Contributors. Maybe easier and more practical to do all that in Slack (we actually talked about that before and agreed on Slack 😅)? And we can decide on a yes-no-neutral basis with the same rules?

Author: I think yes no neutral with same rules is good. I do think the counts of votes for at least proposals should be visible publicly and not just in slack. I'm not sure if each person individual vote should be public though

### Core Contributors
Core Contributors are those who have provided consistent and meaningful contributions to ArviZ.
These can be, but are not limited to, code contributions, community contributions, tutorial

> development etc. Core Contributors will be given the ability to manage the ArviZ github

Reviewer 4: This is only for core contributors?

Reviewer: We have to make sure we follow whatever we decide here before merging. IIUC, the writing above would mean people with merge privileges who is not core contributor has to become core contributor or lose the merge privileges

Author: I agree, we need to make things consistent before we merge this.

Reviewer 2: Yeah, I agree with Oriol's interpretation.

repository, including code merges to master. This does not necessarily mean Core Contributors
must submit code, but more so signifies trust with the project as a whole.

#### Core Contributor Responsibilities
* Enforce code of conduct
* Maintain a check against Council

#### Core Contributor Nominations and Confirmation Process

> Current Core Contributors can nominate candidates for consideration by the council. On the first

Reviewer: I think there should be something about nominated people accepting before triggering the voting process

of the month following a nomination the Council will vote on each nominee where greater than 50% of the votes
means a confirmation. In the event of a tie a second election will occur where all core contributors can vote,

> if 50% or greater of the community confirms the nominee will be accepted.

Reviewer 4: I start here.

What options do we have?

Binary voting
--> majority rules: at least 50 % of the votes
--> minority rukes: at least x % of the votes (x < 50%)
--> ... rules: at least y % of the votes (y > 50%)
--> all agree: 100 % of the votes

What other systems are possible?

Reviewer: I think we should go with yes-no-neutral as we were discussing to do with council elections. In that case acceptance could be decided on a having more yes than no base, or as having yes go over a threshold (without taking into account if the rest of votes are no or neutral), or having two thresholds (i.e. yes larger than X% and no below Y%), etc. @AlexAndorra probably knows better the weakness and strengths of each system

Reviewer 2: As this is a binary choice here (one proposal at a time) and not a multichoice setting (several proposals compete against each other at the same time), we can't use Majority Judgment, but I like the yes-no-neutral and two-threshold process! 
That way, we can ensure that a majority of the community likes the proposed idea _and_ there isn't a substantial part of the community that dislikes the idea. Also, it's the way we elect Council Members in the first round, so it ties things up nicely 😉 
Setting the thresholds is hard and artificial, but what about yes >= 50% and no <= 25% (i.e at least half of community likes and not more than a fourth of it dislikes)?

Author: I am no opposed to idea but have question. Per NumFOCUS's second comment, and critical to this voting style, what happens if both no and yes thresholds are achieved? How do we resolve?

Reviewer 2: Both conditions are necessary and sufficient to pass a proposal: yes >= 50% _and_ no <= 25%. If one of these conditions isn't met -- e.g yes = 60% but no = 30% -- the proposal is refused and must be amended before being put to the vote again. The same goes if no = 10% but yes = 40% of course.

Reviewer: @AlexAndorra I don't think council election section says anything about yes-neutral-no voting nor about the 2 thresholds, it should probably be udated together with some rules to change the number of council members

Author: How do we amend a core contributor add though? That's a pretty binary proposal so if we hit both the yes and no threshold I don't get what the next step is

Reviewer 2: @OriolAbril I think I answered you on Slack.
@canyon289 I'm not sure about what you mean by "amend a core contributor add", but, in your example, if the no threshold is hit, then the proposed core contributor is rejected, even though the yes threshold is hit too -- both conditions must be true (yes >= 60% _and_ no <= 20%) for the core contributor to be accepted.

Author: Core contributor add meant if we were looking to add a core contributor. Ok this makes sense now ( i was just too slow to get it)

> Current Core Contributors can nominate candidates for consideration by the council. On the first
> of the month following a nomination the Council will vote on each nominee where greater than 50% of the votes
> means a confirmation. In the event of a tie a second election will occur where all core contributors can vote,
> if 50% or greater of the community confirms the nominee will be accepted.

Reviewer 2: To follow up on Oriol's suggestion (+ some punctuation nitpicks):
```suggestion
Current Core Contributors can nominate candidates to become Core Contributors. If nominated candidates accept their nomination (explicit comment approving nomination on the GitHub issue or "thumbsup" emoji on the same issue), then they can be considered by the Council: on the first of the month following a nomination, the Council will vote on each nominee, where greater than 50% of the votes means a confirmation. In the event of a tie, a second election will occur, where all core contributors can vote. If 50% or greater of the community confirm, the nominee will be accepted.
```

* Core contributors can also voluntarily leave the project by notifying the community through a public means or by notifying the entire council.

#### Voting Criteria For Future Elections
Voting for first election is restricted to establish stable governance, and to defer major decision to elected leaders
* For the first election only the people registered following the guidelines in `elections/ArviZ_2020.md` can vote
* In the first year, the council must determine voting eligibility for future elections between two criteria:
  * Core contributors
  * The contributing community at large

### Core Contributors
Core Contributors are those who have provided consistent and meaningful contributions to ArviZ.
These can be, but are not limited to, code contributions, community contributions, tutorial
development etc. Core Contributors will be given the ability to manage the ArviZ GitHub
repository, including code merges to master. This does not necessarily mean Core Contributors
must submit code, but more so signifies trust with the project as a whole.

#### Core Contributor Responsibilities
* Enforce code of conduct

> * Maintain a check against Council

Reviewer 3: How do they function as a check against the Council?

Reviewer: I understand that the idea is that the check is generally informal with the vote of no confidence as last resort if discussion and warnings in github and slack do not work
